x/version/version_height_mapping.go

```diff
@@ -0,0 +1,79 @@
+package version
+
+import (
+	"math"
+	"sort"
+)
+
+// ChainVersionConfig stores a set of version ranges and provides a method to
+// retrieve the correct version for a given height.
+type ChainVersionConfig struct {
+	Ranges []HeightRange
+}
+
+// NewChainVersionConfig creates a new ChainVersionConfig from a map of app versions to starting heights.
+func NewChainVersionConfig(versions map[uint64]int64) ChainVersionConfig {
+	return ChainVersionConfig{
+		Ranges: createRange(versions),
+	}
+}
+
+// GetVersion returns the app version for a given height.
+func (v ChainVersionConfig) GetVersion(height int64) (appVersion uint64) {
+	return getVersion(height, v.Ranges)
+}
+
+// HeightRange is a range of heights that a version is valid for. It is an
+// internal struct used to search for the correct version given a height, and
+// should only be created using the createRange function. Heights are
+// non-overlapping and inclusive, meaning that the version is valid for all
+// heights >= Start and <= End.
+type HeightRange struct {
+	Start   int64
+	End     int64
+	Version uint64
+}
+
+// createRange creates a set of heightRange structs from a map of version
+// strings to start heights.
+func createRange(versions map[uint64]int64) []HeightRange {
+	ranges := make([]HeightRange, 0, len(versions))
+	for version, start := range versions {
+		ranges = append(ranges, HeightRange{
+			Version: version,
+			Start:   start,
+		})
+	}
+
+	sort.Slice(ranges, func(i, j int) bool {
+		return ranges[i].Start < ranges[j].Start
+	})
+
+	for i := 0; i < len(ranges)-1; i++ {
+		ranges[i].End = ranges[i+1].Start - 1
+	}
+
+	// set the end of the last range to the max uint64 to cover all heights
+	ranges[len(ranges)-1].End = math.MaxInt64
+
+	return ranges
+}
+
+// getVersion returns the app version for a given block height. It performs a
+// binary search on the ranges slice.
+func getVersion(height int64, ranges []HeightRange) uint64 {
+	// Perform binary search on the ranges slice
```

Comment:
lol, good job but I feel this is probably overkill. We might have 10 - 20 versions in our lifetime of this